cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,327 @@
+        tags from context.tags_to_ignore will have no impact on the routing procedure.
+
+        Args:
+            circuit: the input circuit to be transformed.
+            max_search_radius: the maximum number of times the cost function can be iterated for
+                convergence.
+            preverse_moment_structure: whether or not the transfomer should preserve the given
+                moment structure of 'circuit'.
+            tag_inserted_swaps: whether or not a RoutingSwapTag should be attched to inserted swap
+                operations.
+            initial_mapper: an initial mapping strategy (placement) of logical qubits in the
+                circuit onto physical qubits on the device.
+            context: transformer context storing common configurable options for transformers.
+
+        Returns:
+            The routed circuit executable on the harware with the same unitary as 'circuit'.
+
+        Raises:
+            ValueError: if circuit has operations that act on 3 or more qubits.
+        """
+
+        # 0. Handle CircuitOperations by unrolling them.
+        if context is not None and context.deep is True:
+            circuit = transformer_primitives.unroll_circuit_op(circuit, deep=True)
+        if not all(protocols.num_qubits(op) <= 2 for op in circuit.all_operations()):
+            raise ValueError("Input circuit must only have ops that act on 1 or 2 qubits.")
+
+        # 1. Do the initial mapping of logical to physical qubits.
+        if initial_mapper is None:
+            initial_mapper = line_initial_mapper.LineInitialMapper(self.device_graph)
+        self._initial_mapping = initial_mapper.initial_mapping(circuit)
+
+        # 2. Construct a mapping manager that implicitly keeps track of this mapping and provides
+        # convinience methods over the image of the map on the device graph.
+        self._mm = mapping_manager.MappingManager(self.device_graph, self._initial_mapping)
+
+        # 3. Get timesteps and single-qubit operations.
+        timesteps, single_qubit_ops = self._get_timesteps_and_single_qubit_ops(
+            circuit, preserve_moment_structure
+        )
+
+        # 4. Do the routing and save the routed circuit as a list of moments.
+        routed_ops = self._route(
+            timesteps, single_qubit_ops, max_search_radius, tag_inserted_swaps=tag_inserted_swaps
+        )
+
+        # 5. Return the routed circuit by packing each inner list of ops as densely as posslbe and
+        # preserving outer moment structure.
+        return circuits.Circuit(circuits.Circuit(m) for m in routed_ops)
+
+    def _get_timesteps_and_single_qubit_ops(
+        self, circuit: 'cirq.AbstractCircuit', preserve_moment_structure: bool
+    ) -> Tuple[List[List['cirq.Operation']], List[List['cirq.Operation']]]:
+        """Returns the timesteps of the circuit and the single-qubit operations tagged with the
+        timestep they are to be inserting in."""
```

Comment:
Summary docstring should be a single line followed by a newline and a detailed docstring (if we need a detailed docstring).